src/allmydata/test/test_istorageserver.py

```diff
@@ -1046,18 +1086,80 @@ def disconnect(self):
         assert self.storage_client is not current
 
 
+class _HTTPMixin(_SharedMixin):
+    """Run tests on the HTTP version of ``IStorageServer``."""
+
+    def setUp(self):
+        if PY2:
+            self.skipTest("Not going to bother supporting Python 2")
+        return _SharedMixin.setUp(self)
+
+    def _get_istorage_server(self):
+        set_treq_pool(HTTPConnectionPool(reactor, persistent=False))
+        swissnum = b"1234"
+        self._http_storage_server = HTTPServer(self.server, swissnum)
+        self._port_number = allocate_tcp_port()
```

Comment:
new calls to `allocate_tcp_port` are not allowed.  all tests that use `allocate_tcp_port` have spurious intermittent failures.

:(

*Fortunately* since the call to `listenTCP` is the very next thing this seems easy to fix.  Just listen on 0 and then ask `self._listening_port` what port it got (and put the whole thing in a retry loop that handles the occasional error you get from trying to bind port 0).

A bigger change would be to use `allmydata.test.common.SameProcessStreamEndpointAssigner` (which is like a fixture but not quite).  That's what most of the other tests that want to listen on a real TCP port do.  Apart from dealing with port collisions this makes it easy to change how all those tests work at once.

But you also want to construct a URL and `SameProcessStreamEndpointAssigner` is not set up to help you with that much so I guess it's probably not the right tool to use here... though... the reason it returns its results the way it does is because that's what Foolscap wants to consume.  So if we're eventually going to have all HTTP-based stuff and no Foolscap-based stuff, it will (probably soon?) make a lot of sense to either fix `SameProcessStreamEndpointAssigner` to be HTTP friendly or add something else similar that is.

This is probably worth dealing with sooner rather than later so we can write most of the HTTP-based tests safely.  Maybe it could be a near-term followup after this PR lands (or after you resolve your current branch stack).  Maybe `SameProcessStreamEndpointAssigner` doesn't even reflect the best strategy for HTTP-based tests and it would be good to figure that out soon, too, so we can write most of the new tests in a shape that is amenable to safe server setup instead of having to do a lot of fixing up later on.

Reply:
The location hint could probably be tweaked to be a URL quite easily? So will see if I can use SPSEA.

Reply:
Done.